File: gobang/gobang_c++/AI/comparer.cpp
//
// Created by ycm on 2023/12/9.
//

#include "comparer.h"

bool comparer::equal(double a, double b){
    using namespace comparerConstVar;
    b = (b==0)?infinitesimal:b;
    return ((b>=0)?(b/threshold<=a and a<=b*threshold):(b*threshold<=a and a<=b/threshold));
}

bool comparer::greater(double a, double b){
    using namespace comparerConstVar;
    return (b>=0)?(a>=(b+delta)*threshold):(a>=(b+delta)/threshold);
}

bool comparer::greaterEqual(double a, double b){
    return comparer::equal(a, b) or comparer::greater(a, b);
}

bool comparer::less(double a, double b){
    using namespace comparerConstVar;
    return (b>=0)?(a<=(b-delta)/threshold):(a<=(b-delta)*threshold);
}

bool comparer::lessEqual(double a, double b){
    return comparer::equal(a, b) or comparer::less(a, b);
}

File: gobang/gobang_python/AI/comparer.py
_threshold = 1.15
_delta = 0.1
_infinitesimal = 0.01  # 无穷小


def equal(x, y):
    y = _infinitesimal if y == 0 else y
    return y / _threshold <= x <= y * _threshold if y >= 0 else y * _threshold <= x <= y / _threshold


def greater(x, y):
    return x >= (y + _delta) * _threshold if y >= 0 else x >= (y + _delta) / _threshold


def greaterEqual(x, y):
    return equal(x, y) or greater(x, y)


def less(x, y):
    return x <= (y - _delta) / _threshold if y >= 0 else x <= (y - _delta) * _threshold


def lessEqual(x, y):
    return equal(x, y) or less(x, y)
